### PR TITLE
Allow reading binary doc values as a DataInput

### PR DESCRIPTION
**Pull request by iverase**
see https://github.com/apache/lucene/issues/12459

### CODE REVIEW

File: lucene/core/src/java/org/apache/lucene/codecs/lucene90/Lucene90DocValuesProducer.java [hunk 1]
```diff
@@ -820,6 +822,201 @@ public BytesRef binaryValue() throws IOException {
...
+      return disi.advanceExact(target);
+    }
+  }
+
+  private static class SlicedDataInput extends DataInput {
+
+    private final IndexInput in;
+    private int length;
+
+    private int read;
```
Comment (jpountz) [inline lucene/core/src/java/org/apache/lucene/codecs/lucene90/Lucene90DocValuesProducer.java:903]: Do we need to cache it, or could we check in#getFilePointer on every access?

**Review by stefanvodita — COMMENTED** [review]
Interesting idea! It looks useful. Are you planning to replace some existing uses of `BinaryDocValues` with `DataInputDocValues`? I checked the code for places where it might make more sense to use `DataInputDocValues`:
1. [MatchingFacetSetsCounts](https://github.com/apache/lucene/blob/fb8183332bc5fa74a8eb36105be1026ae7aa0a8e/lucene/facet/src/java/org/apache/lucene/facet/facetset/MatchingFacetSetsCounts.java#L105-L107).
2. [TaxonomyFacetIntAssociations](https://github.com/apache/lucene/blob/fb8183332bc5fa74a8eb36105be1026ae7aa0a8e/lucene/facet/src/java/org/apache/lucene/facet/taxonomy/TaxonomyFacetIntAssociations.java#L75-L82).
3. [TaxonomyFacetFloatAssociations](https://github.com/apache/lucene/blob/fb8183332bc5fa74a8eb36105be1026ae7aa0a8e/lucene/facet/src/java/org/apache/lucene/facet/taxonomy/TaxonomyFacetFloatAssociations.java#L182-L189).
4. [SerializedDVStrategy](https://github.com/apache/lucene/blob/fb8183332bc5fa74a8eb36105be1026ae7aa0a8e/lucene/spatial-extras/src/java/org/apache/lucene/spatial/serialized/SerializedDVStrategy.java#L199-L202).

The last one falls under the `ByteArrayInputStream` case [previously mentioned](https://github.com/apache/lucene/issues/12459#issuecomment-1650744812).

File: lucene/core/src/java/org/apache/lucene/codecs/lucene90/Lucene90DocValuesProducer.java [hunk 2]
```diff
@@ -820,6 +822,213 @@ public BytesRef binaryValue() throws IOException {
...
+    }
+
+    void init(long offset, int length) throws IOException {
+      in.seek(offset);
+      this.length = length;
+      this.offset = offset;
+    }
+
+    private void checkBounds(int numBytes) throws EOFException {
+      if (Math.toIntExact(getPosition() + numBytes) - 1 >= length) {
```
[the remark this answers is not in the document]
Comment (iverase) [reply lucene/core/src/java/org/apache/lucene/codecs/lucene90/Lucene90DocValuesProducer.java]: sure thanks!

**Comment by jpountz** [comment]
The more I think of this change, the more I like it: most of the time, you would need to read data out of binary doc values, e.g. (variable-length) integers, strings, etc. and exposing binary doc values as a data input not only makes this easier to do, but also more efficient by saving one memory copy.

Thoughts:
 - We are losing the symmetry between the index-time API (BinaryDocValuesField), which takes a byte[] while the search-time API produces a `DataInput`. Not a deal breaker, but users have to be careful about endianness.
 - Today it's possible to consume values multiple times, but the data input is stateful, so if one consumer reads data from it, and then another consumer tries to read data again it will fail unless it resets the input via a call to `advanceExact`. Again not a deal breaker but something that could be a surprise to some callers.

In terms of backward compatibility, I'm contemplating not introducing a new `DataInputDocValues` class, and instead have a `dataInput()` method on `BinaryDocValues`, as well as a default implementation of `binaryValue()` that reads all bytes from the data input. What do you think? I'm also curious of the opinion of our sophisticated backward compatibility policeman @uschindler.

To @stefanvodita 's point, it would be nice to migrate at least one or two call sites that would get simpler with a `DataInput` to further validate this change.

**Comment by iverase** [comment]
> I'm contemplating not introducing a new DataInputDocValues class, and instead have a dataInput() method on BinaryDocValues

I think this approach defeats on of the main purposes for this change, that is to avoid allocating a byte array when reading doc values. I don't think we want BinaryDocValues to do that lazily:

```
when one of the doc values is big, in the order of few megabytes, it can cause issues with small heaps (or even big heaps if 
big enough). This is due to the allocation of a big byte array upfront, that can be consider humongous allocations by the G1 
garbage collector and it can cause heap issues under high load.
```

On my own use case, getting a DataInput is not enough as I need random access via get/set position, in a similar fashion to what I am doing now via ByteArrayDataInput.

**Comment by jpountz** [comment]
> I think this approach defeats on of the main purposes for this change, that is to avoid allocating a byte array when reading doc values. I don't think we want BinaryDocValues to do that lazily

What is the problem with allocating lazily? It wouldn't make sense to me with the current API, where binaryValue() is the only way to retrieve the data, but if it were to only remain for bw compat it would make sense to me to only incur the byte[] overhead if the legacy `binaryValue()` API is used?

> On my own use case, getting a DataInput is not enough as I need random access via get/set position, in a similar fashion to what I am doing now via ByteArrayDataInput.

This has been a challenge so many times in the past, maybe it's time to add `seek()` support to `DataInput`?

**Comment by uschindler** [comment]
> This has been a challenge so many times in the past, maybe it's time to add `seek()` support to `DataInput`?

We have full random access (positional reads), if you extend the interface `RandomAccessInput`.

Adding seek support to `DataInput` is not a good idea because it would prevent those `DataInput`s based on `InputStream` (e g. `InputStreamDataInput`, which is used at various places).

**Comment by uschindler** [comment]
To save more memory copies, the codec may use a slice from the underlying IndexInput directly to support both access apis. All file pointer checks would then be performed by the low level JVM intrinsics used my MemorySegmentIndexInput's slices. If you use those views and not let them escape the current method, the GC pressure isn't there (be careful: Profilers show them as escape analysis no longer works when the profiler prevents inlining).

At some point we must anyways go away from the reusing of instances and move Lucene to the shortliving immutable instances used by modern JDK APIs. Escape analysis works fine, if you have simple apis.

**Comment by iverase** [comment]
Thanks @jpountz and @uschindler for the input. I had a look into `RandomAccessInput` and I don't think this what we need. We need an DataInput that is positional ware so it supports seek and in addition it knows its length so we can still read the full stream via `readBytes(byte[] b, int offset, int len)` to support the current functionality.

What I am missing is an abstraction between `DataInput` and `IndexInput` like:

```
/**
 * A positional aware {@link DataInput}.
 */
public abstract class SliceDataInput extends DataInput {

    /** The number of bytes in the stream. */
    public abstract long length();

    /**
     * Returns the current position in this stream, where the next read will occur.
     *
     * @see #seek(long)
     */
    public abstract long position();

    /**
     * Sets current position in this stream, where the next read will occur.
     *
     * @see #position()
     */
    public abstract void seek(long pos) throws IOException;

    /**
     * {@inheritDoc}
     *
     * <p>Behavior is functionally equivalent to seeking to <code>position() + numBytes</code>.
     *
     * @see #position()
     * @see #seek(long)
     */
    @Override
    public void skipBytes(long numBytes) throws IOException {
        if (numBytes < 0) {
            throw new IllegalArgumentException("numBytes must be >= 0, got " + numBytes);
        }
        final long skipTo = position() + numBytes;
        seek(skipTo);
    }
}
```
(Naming is so hard, naming proposals are welcome) 

This abstraction can be extended by `ByteBuffersDataInput` and `IndexInput` and potentially by `ByteArrayDataInput` and `FTS.BytesReader`.

**Comment by iverase** [comment]
I have been thinking a bit longer about this and I think this approach of `DataInput` is not right. Instead we should try to return an API more similar to `RandomAccessInput` as Uwe suggested. 

What convinces me is the fact that `BytesRef` can easily implement (and maybe should) RandomAccessInput. Unfortunately it is not straightforward to change BytesRef with RandomAccessInput as we are doing things with BytesRef that are challenging to implement generically:

- Maybe we should add the length of the `RandomAccessInput` to the API? ` public long length()`
- We can copy the contents of a `RandomAccessInput` byte by byte, should we add a bulk copy method like  ` public byte readBytes(long pos, byte[] b,  int offset, int length) throws IOException;`?
- Can RandomAccessInput implement Comparable?
- Can RandomAccessInput implementation have implementations of equals and hashCode?

The first two are easy to implement and maybe we should do it regardless this issue. The other two are much trickier.

File: lucene/core/src/java/org/apache/lucene/codecs/lucene90/Lucene90DocValuesProducer.java (hunk 1 above)
Comment (iverase) [reply lucene/core/src/java/org/apache/lucene/codecs/lucene90/Lucene90DocValuesProducer.java]: I move the API to RandomAccessInput,

**Comment by iverase** [comment]
I open https://github.com/apache/lucene/pull/13948 which is clearly less invasive.

**Pull request closed without merging**